### PR TITLE
Some bugfixes and additions required to run most of the SYCL-BLAS tests

### PR DESCRIPTION
**Pull request by fodinabor**
This adds some missing builtins (including mul24 #380, although I wasn't willing to make up a concise test for this).

`init_from` only allowed an allocator of the same type.

`sycl::vec`'s load and store have been (naively) implemented.

`multi_ptr`'s assignment operators missed a return statement 😡

With this and the proposed changes in the SYCL-BLAS repo (+ build system adoption in my fork), one can now use SYCL-BLAS with hipSYCL (at least on GPU all unit tests pass, on CPU some tests fail due to illegal memory access in the kernels, which I'd expect to be SYCL-BLAS issues..). For the benchmarks to run, profiling is still missing.
The kernel configurations in my fork are also completely un-tuned..
See my fork for the build system: https://github.com/fodinabor/sycl-blas/tree/feature/hipsycl-build

### CODE REVIEW

File: include/hipSYCL/sycl/libkernel/builtin.hpp [hunk 1]
```diff
@@ -39,11 +39,42 @@ template<class T>
 T __spirv_ocl___hipsycl_min(T a, T b){ return a < b ? a : b;}
 template<class T>
 T __spirv_ocl___hipsycl_max(T a, T b){ return a > b ? a : b;}
+template<class T>
+T __spirv_ocl___hipsycl_mul24(T a, T b){ return a * b;}
```
Comment (illuhad) [inline include/hipSYCL/sycl/libkernel/builtin.hpp]: I think there probably should be `__spirv_ocl_mul24`

File: include/hipSYCL/sycl/libkernel/builtin.hpp [hunk 2]
```diff
@@ -39,11 +39,42 @@ template<class T>
...
+template<class T>
+T __spirv_ocl___hipsycl_abs(T a){ return a < 0 ? -a : a;}
```
Comment (illuhad) [inline include/hipSYCL/sycl/libkernel/builtin.hpp]: Is there no `__spirv_ocl_abs`?

File: include/hipSYCL/sycl/libkernel/builtin.hpp [hunk 3]
```diff
@@ -39,11 +39,42 @@ template<class T>
...
 #else
 template<class T>
+inline HIPSYCL_KERNEL_TARGET
 T __hipsycl_min(T a, T b){ return a < b ? a : b;}
 template<class T>
+inline HIPSYCL_KERNEL_TARGET
 T __hipsycl_max(T a, T b){ return a > b ? a : b;}
+
+#if (defined(HIPSYCL_PLATFORM_HIP) || defined(HIPSYCL_PLATFORM_CUDA)) && defined(SYCL_DEVICE_ONLY)
```
Comment (illuhad) [inline include/hipSYCL/sycl/libkernel/builtin.hpp]: `#if HIPSYCL_LIBKERNEL_IS_DEVICE_PASS_HIP || HIPSYCL_LIBKERNEL_IS_DEVICE_PASS_CUDA` ?

**Comment by fodinabor** [comment]
@hipsycl-ci-bot test please :)

File: include/hipSYCL/sycl/libkernel/math.hpp [hunk 4]
```diff
@@ -117,6 +117,9 @@ namespace sycl {
   HIPSYCL_DEFINE_FLOATN_TRINARY_MATH_FUNCTION(func, func)
 #endif
 
+HIPSYCL_DEFINE_FLOATING_POINT_OVERLOAD(abs, fabsf, fabs)
+HIPSYCL_DEFINE_FLOATN_MATH_FUNCTION(abs, abs)
```
Comment (illuhad) [inline include/hipSYCL/sycl/libkernel/math.hpp]: We already have fabs for floating point types. The spec does not define abs for floats as far as I can see.

**Review by illuhad — APPROVED** [review]

**Pull request merged**